server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java

```diff
@@ -451,18 +454,24 @@ public RoutingAllocation immutableClone() {
         );
     }
 
-    public RoutingAllocation mutableCloneForSimulation() {
+    public RoutingAllocation mutableCloneForSimulation(int targetRelocatingShardsDuringSimulation) {
         return new RoutingAllocation(
             deciders,
             clusterState.mutableRoutingNodes(),
             clusterState,
             clusterInfo,
             shardSizeInfo,
             currentNanoTime,
-            true
+            true,
+            targetRelocatingShardsDuringSimulation
         );
     }
 
+    public boolean hasTargetRelocatingShards() {
```

Comment:
I'm inclined to add boolean return value contracts to the relevant method calls in BalancedShardsAllocator, rather than checking with the RoutingNodes after the fact. I'm concerned that the BalancedShardsAllocator could move multiple shards before we check hasTargetRelocatingShards, instead of building the requirement into the contract of the allocator methods.